dropshot/src/server.rs

```diff
@@ -543,18 +545,83 @@ impl<C: ServerContext> Service<&TlsConn> for ServerConnectionHandler<C> {
     }
 }
 
-/**
- * A running Dropshot HTTP server.
- */
-pub struct HttpServer<C: ServerContext> {
+/// Marker trait used to control whether or not an HTTP server can be closed.
+pub trait Closer: Send + Sync + Unpin + 'static {}
+
+/// Implements [Closer] and identifies that the server can be closed.
+///
+/// This implies:
+/// - [HttpServer::close] can be called
+pub struct Closeable {}
+
+/// Implements [Closer] and identifies that the server cannot be closed
+/// directly.
+pub struct Uncloseable {}
+
+impl Closer for Closeable {}
+impl Closer for Uncloseable {}
+
+/// Marker trait used to control whether or not an HTTP server can be joined.
+pub trait Joiner: Send + Sync + 'static {}
+
+/// Implements [Joiner] and identifies that the server can be joined
+pub struct Joinable {}
+
+/// Implements [Joiner] and identifies that the server cannot be joined
+/// directly.
+pub struct Unjoinable {}
+
+impl Joiner for Joinable {}
+impl Joiner for Unjoinable {}
```

Comment:
TECHNICALLY none of these traits are necessary. I *could* just drop them, and panic if you call "close" or "join" when you shouldn't.

However, I figured it would be cleaner to provide some compile-time guarantees -- you can't `.await` a server if you already called `.get_joiner` on it, and you can't `close` a server if you already called `.get_closer` on it.